common/Geometry2d/Rect.hpp

```diff
@@ -63,19 +63,61 @@ class Rect : public Shape {
 
     Point center() const { return (pt[0] + pt[1]) / 2; }
 
-    /* Assumes that pt[0] <= pt[1] for both x and y */
+    /*
+    * The expand function will make the rectangle larger to include
+    * the given point (just large enough)
+    */
     void expand(Point pt);
+    
+    /*
+    * The expand function will make the rectangle larger to include
+    * the given rectangle (just large enough)
+    */
     void expand(const Rect& rect);
 
+    /*
+    * Makes the rectangle bigger in all direction by the padding amount
+    * especially useful around the goalboxes to movement
+    */
+    void pad(float padding);
+
     float minx() const { return std::min(pt[0].x(), pt[1].x()); }
     float miny() const { return std::min(pt[0].y(), pt[1].y()); }
     float maxx() const { return std::max(pt[0].x(), pt[1].x()); }
     float maxy() const { return std::max(pt[0].y(), pt[1].y()); }
 
+    /*
+    * The corners() function lists the 4 corners of the rectangle
+    * in a predictable order regardless of the 2 corners defined on
+    * construction.
+    * BottomLeft, TopLeft, TopRight, BottomRight
+    * exposed to python as corners()
+    */
+    std::vector<Point> corners();
+
+    /*
+    * The pointList() function is primarily for debugging and 
+    * lists the 2 corner points used to define this rectangle
+    * useful since any 2 diagonal corners can define a rect.
+    */
+    std::vector<Point> pointList();
+
     bool nearPoint(Point pt, float threshold) const override;
     bool nearSegment(const Segment& seg, float threshold) const;
 
     bool intersects(const Rect& other) const;
+    
+    /*
+    * Calculates intersection point(s) between the rectangle and a line segment
+    * Uses Cohen Sutherland line clipping algorithm
+    */
+    bool intersects_(const Segment& other, Point* intr1, Point* intr2) const;
```

Comment:
This can just be named `intersects` and function overloading will determine which one they reach

Reply:
frankly... I just didn't want to deal with it in boost- how big a deal is this?

Reply:
I think it's a pretty big deal, but not earth shattering. It would mess up the function overloading in C++ land. I made a patch for it since I was super familiar with how to do it. I don't think boost allows overloading for python functions so you have to build a second function, that calls the overloaded function.

Here's a link to the [patch](https://gist.github.com/JNeiger/96f2fa64c92cd0736d757dd4bd240122). Basically download the zip. Unzip it into the `robocup-software` directory. Run `git apply fix_underscore.patch` to apply the patch. You'll have to commit this afterwards then.

Feel free to ignore this and try it yourself if you would like, but it sounds like you've fully experienced all the boost::python you wanted to :P

Reply:
ha- if you can't tell from the degradation of my commit messages, I think I'm done with C++ now and want to get back to the data driven stuff we've discussed.... trying so hard to close this out.  But I'll refactor this.

Reply:
done